src/session/content/message_row/photo.rs

```diff
@@ -0,0 +1,160 @@
+use glib::clone;
+use gtk::{gdk, gio, glib, prelude::*, subclass::prelude::*, CompositeTemplate};
+use tdgrand::{enums::MessageContent, types::File};
+
+use crate::session::chat::{BoxedMessageContent, Message};
+use crate::session::content::{message_row::Media, MessageRow, MessageRowExt};
+use crate::utils::parse_formatted_text;
+use crate::Session;
+
+mod imp {
+    use super::*;
+    use glib::WeakRef;
+    use std::cell::RefCell;
+
+    #[derive(Debug, Default, CompositeTemplate)]
+    #[template(resource = "/com/github/melix99/telegrand/ui/content-message-photo.ui")]
+    pub struct MessagePhoto {
+        pub binding: RefCell<Option<gtk::ExpressionWatch>>,
+        pub handler_id: RefCell<Option<glib::SignalHandlerId>>,
+        pub old_message: WeakRef<glib::Object>,
+        #[template_child]
+        pub media: TemplateChild<Media>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for MessagePhoto {
+        const NAME: &'static str = "ContentMessagePhoto";
+        type Type = super::MessagePhoto;
+        type ParentType = MessageRow;
+
+        fn class_init(klass: &mut Self::Class) {
+            Self::bind_template(klass);
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for MessagePhoto {
+        fn constructed(&self, obj: &Self::Type) {
+            self.parent_constructed(obj);
+            obj.connect_message_notify(|obj, _| obj.update_widget());
+        }
+    }
+
+    impl WidgetImpl for MessagePhoto {}
+}
+
+glib::wrapper! {
+    pub struct MessagePhoto(ObjectSubclass<imp::MessagePhoto>)
+        @extends gtk::Widget, MessageRow;
+}
+
+impl MessagePhoto {
+    pub fn new(message: &glib::Object) -> Self {
+        glib::Object::new(&[("message", message)]).expect("Failed to create MessagePhoto")
+    }
+
+    fn update_widget(&self) {
+        let self_ = imp::MessagePhoto::from_instance(self);
+
+        if let Some(binding) = self_.binding.take() {
+            binding.unwatch();
+        }
+
+        if let Some(old_message) = self_.old_message.upgrade() {
+            if let Some(id) = self_.handler_id.take() {
+                old_message.disconnect(id);
+            }
+        }
```

Comment:
Here is just one check necessary, as the others are guaranteed to be not 'None'.
```suggestion
        if let Some(binding) = self_.binding.take() {
            binding.unwatch();
            self_
                .old_message
                .upgrade()
                .unwrap()
                .disconnect(self_.handler_id.take().unwrap());
        }
```

Reply:
Maybe I am wrong. The risk is that the WeakRef is not valid anymore.

Reply:
Maybe I was wrong. There is risk that the WeakRef isn't valid any longer.